src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -93,6 +93,32 @@ def make_install_req_from_dist(dist, parent):
     return ireq
 
 
+class DummyCandidate(Candidate):
+    """A candidate that will never be installed (for constraints)"""
+    def __init__(self):
+        # type: () -> None
+        self._name = "<Dummy>"
+        self._version = Version("0")
```

Comment:
😮 Would it work to raise in the version property instead?

Reply:
Er, maybe? Does it matter, though? I might be missing your point here...

Reply:
It seems awkward to me this class is having a property that’s never called, but returns an arbitrary placeholder value. It would be more obvious to raise in `version` instead to make it clear that we don’t plan to call it, and it’s a clear error if we ever do.

I think an even better way to do this is to re-work the class hierarchy, and let type checking handle this. Maybe `DummyCandidate` should inherit from something else that does not have `version` instead, or maybe we should remove `version` from `Candidate`. I’m not sure yet, so it’s likely a better idea to merge this first, and work on refactoring later. (I do want to try some of this.)

Reply:
Ah, I get what you mean. We need a dummy name, because the resolver uses name to identify things, but version is unused.

I'll switch to a propery that asserts (assuming that mypy doesn't whine at me).